### PR TITLE
Fixes #15711 - Redesigned host info to providers.

### PR DESCRIPTION
**Pull request by ShimShtein**
Now you can add your own info by registering a host method that will return info hash.
#### Main files

`app/models/concerns/host_info.rb`: Core concern, responsible for registration and querying of different info providers.
`app/models/concerns/host_info_providers/static_info.rb`: Provider of the basic (and static) information, like hostgroup, interfaces e.t.c.
`app/models/concerns/host_info_providers/puppet_info.rb`: Provider for the puppet specific information, like smart class parameters and smart variables.
`app/models/concerns/host_params.rb`: Became a provider of (global) host parameters.
#### Classification

Removed the hierarchy from `Classification` and left the base class as a service class that knows how to calculate values for a given set of keys (the type of keys does not matter).
The logic that was stored in the child classes moved to PuppetInfo concern.

### CODE REVIEW

**Comment by ShimShtein** [comment]
Fixed the commit message.

File: app/helpers/hosts_helper.rb [hunk 1]
```diff
@@ -42,8 +42,8 @@ def nic_info_js(compute_resource)
   def value_hash_cache(host)
     @value_hash_cache ||= {}
     @value_hash_cache[host.id] ||= begin
-      Classification::ClassParam.new(:host=>host).inherited_values.
-        merge(Classification::GlobalParam.new(:host=>host).inherited_values)
+      host.inherited_puppetclass_parameters.
+        merge(host.inherited_smart_variables)
```
Comment (dLobatog) [inline app/helpers/hosts_helper.rb]: Isn't this missing global params now?

File: app/models/concerns/host_info.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module HostInfo
+  extend ActiveSupport::Concern
+
+  def info
+    info_hash = {}
+
+    self.class.info_providers.each do |provider_method|
```
Comment (dLobatog) [inline app/models/concerns/host_info.rb]: Neat, I suppose this is meant for plugins to add info to the ENC through `register_info_provider`. Can you write a bit of docs about these methods? Something like https://github.com/theforeman/foreman/blob/develop/app/services/foreman/parameter_filter.rb - just a way for people to know what this interface is for

Comment (dLobatog) [inline app/models/concerns/host_info.rb]: Also I might be missing something here but is this about the ENC or is this about something else?  `.info` has been used just for the ENC for a long time,  this seems to add more stuff to it, which is only going to be used by Puppet but makes it look like `.info` is used for other stuff. Is that the case?

File: app/models/concerns/host_info_providers/puppet_info.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+      info_hash = {}
+      info_hash['classes'] = classes_info_hash
+      info_hash['parameters'] = param
+      info_hash['environment'] = param["foreman_env"] if Setting["enc_environment"] && param["foreman_env"]
+
+      info_hash
+    end
+
+    def parameters_info_hash
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/puppet_info.rb]: nitpick, how about calling it `puppet_static_info` ?

File: app/models/concerns/host_info_providers/puppet_info.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      # Static parameters
+      param = parameters_info_hash
+
+      # Parse ERB values contained in the parameters
+      param = SafeRender.new(:variables => { :host => self }).parse(param)
+
+      info_hash = {}
+      info_hash['classes'] = classes_info_hash
+      info_hash['parameters'] = param
+      info_hash['environment'] = param["foreman_env"] if Setting["enc_environment"] && param["foreman_env"]
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/puppet_info.rb]: `param["foreman_env"].present?` instead would work better, `foreman_env' = ' ' right now returns true

File: app/models/concerns/host_info_providers/puppet_info.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+      info_hash
+    end
+
+    def parameters_info_hash
+      params = {}
+      # maybe these should be moved to the common parameters, leaving them in for now
+      params["puppetmaster"] = puppetmaster
+      params["puppet_ca"]    = puppet_ca_server if SETTINGS[:unattended] && puppetca_exists?
+
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/puppet_info.rb]: nitpick, weird empty line here

File: app/models/concerns/host_info_providers/puppet_info.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        params["ip6"] = ip6
+        params["mac"] = mac
+      end
+
+      params.update(smart_variables)
+    end
+
+    def classes_info_hash
+      return [] if self.environment.nil?
+
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/puppet_info.rb]: nitpick, 2 weird empty lines in this method

File: app/models/concerns/host_info_providers/puppet_info.rb [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    def puppetclass_parameters
+      classifier = Classification::Base.new self
+      keys = PuppetclassLookupKey.includes(:environment_classes).parameters_for_class(puppetclass_ids, environment_id)
+      key_hash = hashed_class_keys(keys)
+      values = classifier.values_hash(keys)
+
+      klasses = {}
+      classes.each do |klass|
+        klasses[klass.name] = smart_class_params_for klass, key_hash, values, classifier
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/puppet_info.rb]: nitpick, try to not use parenthesis for well-known APIs (Ruby, Rails & little else)  and use them otherwise, I don't know if you follow another convention or it's random but it gets a bit difficult to read

File: app/services/classification/class_param.rb [hunk 8]
```diff
@@ -1,41 +0,0 @@
-module Classification
```
Comment (dLobatog) [inline app/services/classification/class_param.rb:1]: Kinda confused where did this code go exactly? Can you provide a bit of background?

File: app/models/concerns/host_info_providers/static_info.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    included do
+      register_info_provider :static_info
+    end
+
+    def static_info
+      # Static parameters
+      param = {}
+      # maybe these should be moved to the common parameters, leaving them in for now
+      param["realm"] = realm.name unless realm.nil?
```
Comment (dLobatog) [inline app/models/concerns/host_info_providers/static_info.rb]: Use `.blank?` everywhere, it leaves less room for error

**Comment by dLobatog** [comment]
Thanks @ShimShtein , I'll review again tomorrow - running out of time now, feel free to leave it as it stands now

File: app/helpers/hosts_helper.rb (hunk 1 above)
Comment (ShimShtein) [inline app/helpers/hosts_helper.rb]: No, `Classification::GlobalParam` [was dealing](https://github.com/theforeman/foreman/blob/develop/app/services/classification/global_param.rb) only with smart variables.

File: app/models/concerns/host_info.rb (hunk 2 above)
Comment (ShimShtein) [inline app/models/concerns/host_info.rb]: > Also I might be missing something here but is this about the ENC or is this about something else?

ENC is a Puppet term, but since we want to be configuration platform agnostic (we don't want to be locked to Puppet, but to be able to use other config tools like Ansible) we want to be able to expose our parameters to other systems, for example Ansible's Inventory.
Basically it means that `host.info` should describe the host in an abstract way - not something that is specific to puppet/chef/ansible or whatever, which it already does.

File: app/services/classification/class_param.rb (hunk 8 above)
Comment (ShimShtein) [inline app/services/classification/class_param.rb:1]: Let's start from the background:
I want to separate puppet from the core in Foreman. This PR started as a way of getting rid of references to [puppet properties](https://github.com/theforeman/foreman/blob/develop/app/services/classification/base.rb#L3) from `Classification::Base`.
I started reviewing the class, and concluded that inheritance was not the right solution for multiple key types handling here. So I started to get up the stack and look where the `Classification` classes are used, and found that it was the `#info` method. So I refactored it to be extendable and created two extensions that actually used this extension point: one for the static properties provided by the core (`static_info.rb`) and one for puppet properties and parameters (`puppet_info.rb`). The latter also received the functionality yanked form the `Classification` subclasses.

Now about the `Classification` refactoring: `Classification::Base` is a service class that has one main and two secondary methods. The main method is `#values_hash` that receives a set of keys (a scope, to be more accurate) and returns values for each key in the input that are relevant to a single host. Previously the child classes were used to supply the set of keys based on different properties in the host, but now the invocation of the service moved to `puppetclass_parameters` and `smart_variables` methods in `puppet_info.rb`. These methods are responsible to get the relevant keys, pass them to the `Classification::Base` service and format the result. Two secondary services that are served by the `Classification::Base` are `#inherited_values` which returns a subset of `values_hash` and `#value_of_key` which formats the value for a given key. The `#value_of_key` method is a bit interesting, since it uses a cached `SafeRender` instance, that we don't want to create for each value, and that's the main reason that `Classification::Base` class should be instantiated.

**Comment by ShimShtein** [comment]
Fixed all the nitpicks and answered all comments.

**Comment by ShimShtein** [comment]
Rebased

File: app/models/concerns/host_params.rb [hunk 10]
```diff
@@ -8,59 +8,65 @@ module HostParams
...
 
-      if SETTINGS[:locations_enabled] && location
-        params += extract_params_from_object_ancestors(location)
-      end
+  def host_params
+    return cached_host_params unless cached_host_params.blank?
+    hp = host_inherited_params
+    # and now read host parameters, override if required
+    host_parameters.authorized(:view_params).each { |p| hp[p.name] = p.value }
+    @cached_host_params = hp
```
Comment (dLobatog) [inline app/models/concerns/host_params.rb]: Can we use `Rails.cache` here instead? It'd work better for any deployment where there's more than one host & cache would be more persistent. I'm not sure if 'in-memory' caches such as this one survive when passenger kills it's workers (like during low-activity periods)

File: app/models/concerns/host_params.rb [hunk 11]
```diff
@@ -8,59 +8,65 @@ module HostParams
...
 
-      params += domain.domain_parameters.authorized(:view_params) if domain
-      params += subnet.subnet_parameters.authorized(:view_params) if subnet
-      params += subnet6.subnet_parameters.authorized(:view_params) if subnet6
-      params += operatingsystem.os_parameters.authorized(:view_params) if operatingsystem
-      params += extract_params_from_object_ancestors(hostgroup) if hostgroup
-      params
+  def host_inherited_params_objects
+    params = CommonParameter.all
+    if SETTINGS[:organizations_enabled] && organization
```
Comment (dLobatog) [inline app/models/concerns/host_params.rb]: Nitpick, 

``` ruby
Taxonomy.enabled_taxonomies do |taxonomy_type|
  next unless send(taxonomy_type.singularize).present?
  params += extract_params_from_object_ancestors(send(taxonomy_type.singularize))
end
```

File: app/services/classification/class_param.rb (hunk 8 above)
Comment (dLobatog) [inline app/services/classification/class_param.rb:1]: Thanks, can you add some of that info on top of `values_hash`, `inherited_values`, `possible_value_orders` and `value_of_key`? I think it makes sense now, it's just not that obvious at first glance what they do so if the next reader has an easier time figuring it out, all the better :smile:

**Comment by dLobatog** [comment]
Awesome, thanks @ShimShtein ! In general it looks great, I'm only lacking some docs in the 'new' `Classification::Base` and tests for the `host_info_providers`. 

All in all I think it's a nice refactor and makes it easier to understand - I'm worried that we have that a very hard dependency on the host object (info_providers are concerns) that it was not that strict when those were services. Since I don't see why / how the ENC would be used from a context where there's no host, I guess it's fine, although I'd like more eyes on it. 

Lastly, I can see how making the host_info_providers decouples Puppet from the `Classification` stuff, however, is there anything that's not Puppet that uses the `Classification` service? What I'm trying to get to is why did you not decide to completely separate `info` or `Classification` from Host and make it 'optional' vs this?

Thanks again, good work!

File: app/models/concerns/host_info.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+# module MyInfoProvider
+#   extend ActiveSupport::Concern
+#
+#   def my_info
+#     { 'my_prop1' => 'my_val' }
+#   end
+#
+#   included do
+#     register_info_provider :my_info
+#   end
```
Comment (ares) [inline app/models/concerns/host_info.rb]: Could you please document this example to [redmine wiki](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin) when this is merged? Also could you extend [Plugin DSL](https://github.com/theforeman/foreman/blob/12612809c5deb885b13ceaf36b6a147c21688eb5/app/services/foreman/plugin.rb#L273) so plugin authors can register their info providers easily from initializer?

File: app/models/concerns/host_params.rb (hunk 10 above)
Comment (ShimShtein) [inline app/models/concerns/host_params.rb]: It's a bit risky, since it would be shared between instances. For example look at the use case of changing the hostgroup: first you will get an existing instance of the host (to show it in the UI) the cache is created (with values for the old hostgroup); now the user changes the hostgroup and the AJAX call is performed - a new host instance is created, but if the cached values will be used it will be wrong.

**Comment by ShimShtein** [comment]
Addressed all the issues and redesigned a bit the `Classification::Base`: Now the main method `values_hash` is a class method, and it returns an instance of pre-initialized results object that is responsible for parsing the `LookupValue` objects (basically it's what `value_of_key` did in the past).

@dLobatog 

> I'm worried that we have that a very hard dependency on the host object (info_providers are concerns) that it was not that strict when those were services.

The main dependency is while selecting the keys, which was there before the refactor and I am not sure we can remove that dependency, since the set of keys is determined by the host object. The same dependency was there before the refactoring (it was a little less explicit, because it was delegating some of the methods to host). 

@ares 

> could you extend Plugin DSL so plugin authors can register their info providers easily from initializer?

I am not sure that it is a good idea, since the info provider method should be a method on the host object. I prefer that extensions would register their method inside the concern, just like I do in `puppet_info.rb`.

**Comment by ares** [comment]
> > could you extend Plugin DSL so plugin authors can register their info providers easily from initializer?
> 
> I am not sure that it is a good idea, since the info provider method should be a method on the host object. I prefer that extensions would register their method inside the concern, just like I do in puppet_info.rb.

Since you brought this up, I'd actually prefer providers to be a separate classes, I don't like mixing too much logic into a single object. Anyway if this is too big change without big enough benefit for you, I'd still appreciate a method that would basically only include the concern into the host class. The registration would then happen inside the `included` callback.

My main concern is that we already have this concept for other extendable parts with few minor differences and I'd like to avoid introducing new one, I don't consider this minor difference.

**Comment by ShimShtein** [comment]
> I'd actually prefer providers to be a separate classes, I don't like mixing to much logic into one object.
1. This class should be aware of the host - it's main purpose is to provide set of keys based on some properties of the host.
2. A provider is pure function - it doesn't hold any state, so I don't see why we should encourage creation of classes in this case.

> My main concern is that we already have this concept for other extendable parts with few minor differences

I see this much like callback subscription (actually I was considering usage of callbacks, but they don't support return values). I don't think it would confuse somebody to learn how to use it.

>  I'd still appreciate a method that would basically only include the concern into the host class. 

Will do.

Basically I am a bit concerned about exposing too much information up-front in the plugin registration function. I am afraid that it will become a huge pile of different method enumerators that are not connected to each other.
Very crude example of the problem:

``` ruby
register_plugin :my_plugin do
  extend_something_in_host_model :my_host_extension
  extend_some_other_functionality_in_host :my_other_extension
  extend_host_view :my_view_extension
  extend_some_other_view :my_other_view_extension
end
```

Basically we created logical references between this code to four different places, each dealing with different object and different part of MVC.

File: app/models/concerns/host_info_providers/static_info.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    def add_domain_params(param)
+      return if domain.blank?
+      param["domainname"] = domain.name unless domain.name.blank?
+      param["foreman_domain_description"] = domain.fullname unless domain.fullname.blank?
+    end
+
+    def add_taxonomy_params(param)
+      if SETTINGS[:locations_enabled]
+        param["location"] = location.title unless location.blank?
```
Comment (timogoebel) [inline app/models/concerns/host_info_providers/static_info.rb]: this now needs to look a little different, please see b96e021026cdcc6918c61d246dbf5c7baabfc1b4

**Comment by ares** [comment]
I don't want to go off-topic too much so just short reactions.

> This class should be aware of the host - it's main purpose is to provide set of keys based on some properties of the host.

Yes, but that does not mean it must be defined in that host. You can create provider class that needs host to exist (host specified in it's constructor).

> A provider is pure function - it doesn't hold any state, so I don't see why we should encourage creation of classes in this case.

I know we are far from SRP but I consider adding more responsibilities to a single class as a bad practice. The more we extract from host object the better IMHO.

> I am afraid that it will become a huge pile of different method enumerators that are not connected to each other.

Yes, we're already there and I hope one day we'll unify all of these methods. But I think it's better to add one "similar" API rather then encourage plugin devs to use internal objects.

**Comment by ShimShtein** [comment]
@timogoebel , changed the taxonomy params accordingly, thanks!

@ares, Added a simple method that enables us to add concerns to a host.

If we want a deeper redesign of parameter related methods in host (service class extraction), we should probably open a new RFC, because it will be pretty incompatible with the current design.
I would suggest a proper RFC for dealing with plugin registration mess too. We can discuss the requirements and a rough solution on IRC/Hangout, and then I will be glad to put it on paper for community comments.

File: app/services/foreman/plugin.rb [hunk 14]
```diff
@@ -351,6 +351,11 @@ def parameter_filters(klass)
       @parameter_filters.fetch(klass.is_a?(Class) ? klass.name : klass, [])
     end
 
+    # Add concerns to a host
+    def extend_host(concern_module)
```
Comment (dLobatog) [inline app/services/foreman/plugin.rb]: Bad idea, what does this give the plugin developer? They can just rely on Rails for that...

`https://github.com/theforeman/foreman_ansible/blob/master/lib/foreman_ansible/engine.rb#L51`

File: app/models/concerns/host_params.rb (hunk 10 above)
Comment (dLobatog) [inline app/models/concerns/host_params.rb]: Sorry, I don't follow that logic. How is that different now? In the current case:
1. User gets `@host` in the UI - cache is created if we call host_params 
2. User changes hostgroup -> controller calls `refresh_host` which fetches params from cache - values are wrong.
3. They are _not_ wrong because the `@host` object is regenerated.. but we created a nearly useless cache.

In the case I propose:
1. User gets `@host` in the UI - cache is created if we call host_params. Let's save it as `Rails.cache.write("host_params_#{host.fqdn}", host_params)'.
2. User changes hostgroup -> controller calls `refresh_host` which fetches params from cache - values are wrong. At this point (https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L733) we should call `@host.clear_host_parameters_cache!` and empty the cache. Only after that and after actually submitting changes to a host we should call that method (I don't know why that's not the case now, probably broken)
3. They are _not_ wrong in any case now, no matter any instance they hit, and they're reusable. 

Anyway, I'm realizing it's maybe too much work to do in this PR, give me a nudge at some point and I'll do it separately

File: test/unit/classification_test.rb [hunk 15]
```diff
@@ -314,8 +302,9 @@ def setup
                           :value => "345",
                           :use_puppet_default => true
     end
-    enc = classification.enc
-    refute enc['base'].has_key?(key.key)
+
+    enc = @host.puppetclass_parameters
+    refute enc['base'].key?(key.key)
```
Comment (dLobatog) [inline test/unit/classification_test.rb]: Saw a few of this `has_key` to `key` changes, you may want to do it everywhere via rubocop (in another PR)

File: app/services/foreman/plugin.rb (hunk 14 above)
Comment (ares) [inline app/services/foreman/plugin.rb]: We don't want plugin devs to touch internal Foreman objects, later we might just change how extend_host works without requiring plugin changes, imagine Host::Managed would change to Host::Base e.g.

**Comment by dLobatog** [comment]
Honestly I prefer to discuss it briefly then implement a trivial part of it as a service object than opening an RFC for that. In my opinion RFCs can quickly end up as design by committee and they just delay talking about an implementation - furthermore, they allow us to talk much without having any real code to look at. A _simple_, _small_ proof of concept is more conducive to a good solution in my experience.

File: app/models/concerns/host_params.rb (hunk 10 above)
Comment (ShimShtein) [inline app/models/concerns/host_params.rb]: You are right with the analysis. My consideration is that in step 2 the host object is still in-memory only (if the user cancels the change or just closes the window - the change is not applied), which means that the cache shouldn't be updated.
I can see only two cases when the cache would be effective:
1. For multiple access to parameters on the same instance
2. Read-only access to the host from multiple requests.
Anyway this cache is there for too long (I have found the [commit](https://github.com/theforeman/foreman/commit/8ea536e3cf5cb9566953dc1772973160a99eb41f) where it is added). Maybe it's time to get rid of it... If you want, I'll open a new issue in Redmine to remind us to do that.

File: app/services/foreman/plugin.rb (hunk 14 above)
Comment (ShimShtein) [inline app/services/foreman/plugin.rb]: I added it in response to [Marek's request](https://github.com/theforeman/foreman/pull/3701#issuecomment-238601625).

File: test/unit/classification_test.rb (hunk 15 above)
Comment (ShimShtein) [inline test/unit/classification_test.rb]: We have it only in tests, otherwise Rubocop would complain about it in the build process. Since I touched this file, I have changed it here too.

File: app/services/foreman/plugin.rb (hunk 14 above)
Comment (dLobatog) [inline app/services/foreman/plugin.rb]: I saw the request - I just think it's best to leave this kind of things to Rails. I don't necessarily see it as a bad thing that plugin devs know about objects in Foreman, in fact they pretty much have to know about them. Anyway, not that out of bounds that I would block this PR for that, I'll probably just rely on Rails for extending core models & that's it

Comment (ares) [inline app/services/foreman/plugin.rb]: I don't say plugin devs should not know about internal objects, I don't want them to use them directly. Otherwise plugin gets broken without us knowing, leading to plugin devs frustration etc. I find this very important, plugins need more stable API internal that they can use.

**Comment by ShimShtein** [comment]
Rebased the PR and fixed it according to the RFC: https://github.com/theforeman/rfcs/pull/12

**Comment by ShimShtein** [comment]
I have moved the code with minimal changes from other places, so code climate complains about this code. I thought that it would be easier to review it if the code will retain similarity to the old one. Once you are OK with the high level, I will try to refactor it according to codeclimate's standard.

**Comment by ShimShtein** [comment]
Rebased and redesigned to be more codeclimate friendly.

**Comment by ShimShtein** [comment]
Typo, sorry.

**Comment by ShimShtein** [comment]
Rebased

**Comment by ShimShtein** [comment]
Rebased.

File: app/models/concerns/host_info_extensions.rb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    HostInfo.register_info_provider(HostInfoProviders::ConfigGroupsInfo)
+    HostInfo.register_info_provider(HostInfoProviders::PuppetInfo)
+    HostInfo.register_info_provider(HostInfoProviders::HostParamsInfo)
+  end
+
+  # This is the base method that gathers information from all providers.
+  def info
+    renderer_regex = /renderer\.rb.*host_enc/
+    unless caller.first.match(renderer_regex) || caller[1].match(renderer_regex)
+      Foreman::Deprecation.renderer_deprecation('1.17', __method__, 'host_enc')
```
Comment (ares) [inline app/models/concerns/host_info_extensions.rb:27]: this needs to be 1.18 now

**Review by ares — CHANGES_REQUESTED** [review]
I think this will need changes on Katello side too since they extend ENC output, do we have PR ready for it?

**Comment by timogoebel** [comment]
> I think this will need changes on Katello side too since they extend ENC output, do we have PR ready for it?

I don't want to further delay this PR and didn't read all comments, but: How do plugins in general extend this?

**Comment by ShimShtein** [comment]
@timogoebel, I do this [here](https://github.com/theforeman/foreman/pull/3701/files#diff-6b3be3f7bb2907cfa70c8e541edaaec0R17).

TL;DR:

``` ruby
HostInfo.register_info_provider(HostInfoProviders::PuppetInfo)
```

**Comment by timogoebel** [comment]
> TL;DR:

@ShimShtein : 👍  Got it, thanks.

**Comment by ares** [comment]
> How do plugins in general extend this?

after we get this in, I hope it will be covered in [plugin writing wiki](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin) with some example :-)

**Comment by ShimShtein** [comment]
Rebased and improved registration to match that of facets (should survive reloads better).
So @timogoebel, now plugins can declare:
``` ruby
Foreman::Plugin.register :my_plugin do
  register_info_provider MyPlugin::InfoProvider
end
```

File: app/models/host_info_providers/static_info.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+      param["root_pw"] = host.root_pass unless (!host.operatingsystem.blank? && host.operatingsystem.password_hash == 'Base64')
+    end
+
+    def add_login_params(param)
+      owner = host.owner
+      return unless SETTINGS[:login] && owner
+
+      param["owner_name"]  = owner.name
+      param["owner_email"] = owner.is_a?(User) ? owner.mail : owner.users.map(&:mail)
```
Comment (timogoebel) [inline app/models/host_info_providers/static_info.rb:35]: I believe, this got lost during the rebase:

param["ssh_authorized_keys"] = ssh_authorized_keys
param["foreman_users"] = owner.to_export

**Review by timogoebel — CHANGES_REQUESTED** [review]

**Comment by ShimShtein** [comment]
Went through all params again and added/fixed missing ones. Nice catch @timogoebel, thanks!

File: app/models/host_info_providers/static_info.rb (hunk 17 above)
Comment (timogoebel) [reply app/models/host_info_providers/static_info.rb:35]: Fixed.

**Review by timogoebel — APPROVED** [review]

File: app/models/concerns/host_info_extensions.rb (hunk 16 above)
Comment (ares) [reply app/models/concerns/host_info_extensions.rb:27]: My apologies I messed up, it should have remained 1.17. The deprecation might be removed completely soon but we should probably keep it same in this PR.

File: app/registries/foreman/plugin.rb [hunk 18]
```diff
@@ -416,6 +421,10 @@ def action_scopes_hash_for(controller_class)
       @controller_action_scopes[controller_class.name] || {}
     end
 
+    def host_info_provider(provider_class)
```
Comment (ares) [inline app/registries/foreman/plugin.rb]: when is this supposed to be used? I though the registration is handled by `register_info_provider`

File: app/registries/foreman/plugin.rb [hunk 19]
```diff
@@ -372,6 +372,11 @@ def register_facet(klass, name, &block)
       @facets << Facets.register(klass, name, &block)
     end
 
+    def register_info_provider(klass)
+      @info_providers ||= []
+      @info_providers << HostInfo.register_info_provider(klass)
```
Comment (ares) [inline app/registries/foreman/plugin.rb]: why calling `HostInfo.register_info_provider(klass)`? I thought this plugin instance is keeping the list and HostInfo asks it  here https://github.com/theforeman/foreman/pull/3701/files#diff-adbf4db3b5ffb094ecec3a7a4f1f841bR12

Comment (ShimShtein) [reply app/registries/foreman/plugin.rb]: Current registry scheme enables registering providers from plugins and from arbitrary places in the code (initializers for example).
The concept is that `HostInfo` caches providers from plugins on first access and then you can add additional providers. If the cache is already committed, we need to add the providers directly, and save them for later cache refresh.

**Comment by ShimShtein** [comment]
Removed redundant method from `plugin.rb`

File: app/models/concerns/host_info_extensions.rb (hunk 16 above)
Comment (ShimShtein) [reply app/models/concerns/host_info_extensions.rb:27]: reverted

File: app/registries/foreman/plugin.rb (hunk 19 above)
Comment (ares) [reply app/registries/foreman/plugin.rb]: I'd recommend avoiding the cache in `HostInfo` then. I think providers should be only cached in plugin instance and `HostInfo` should always do local + plugins. Summing few arrays should be fast enough.

the method should look like
```ruby
def register_info_provider(klass)
  @info_providers ||= []
  @info_providers << klass
  klass
end
```
and the HostInfo part:

```ruby
module HostInfo
  class << self
    def providers
      (@providers || []) + entries_from_plugins
    end

    def register_info_provider(provider)
      @providers ||= []
      @providers << provider
      provider
    end
      ...
  end
end
```

**Comment by ShimShtein** [comment]
Removed the caching from HostInfo registry.

**Comment by ares** [comment]
@ShimShtein now the foreman tests looks good, unfortunately another rebase is required. I'm afk next week, if you could schedule the rebase around June 12, I will make my best to finally get this in. I hope the rebase won't be too bad this time.

**Review by dLobatog — CHANGES_REQUESTED** [review]
Moving to WoC because of the rebase required

**Comment by ShimShtein** [comment]
Rebased.

**Comment by ShimShtein** [comment]
@dLobatog, @ares everything looks OK now, care taking a final look?

**Review by ares — APPROVED** [review]

**Comment by ares** [comment]
Works great, diff does not show any missed info in ENC. I tested with Katello PR and it works fine. Merged as 60cf5a375ca96eaf8a7b8b16b8c3cb817186c8a4

**Comment by ares** [comment]
@ShimShtein could you please update the plugin writing wiki?

**Comment by ShimShtein** [comment]
Wiki updated: [diff](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin/diff?utf8=%E2%9C%93&version=118&version_from=117&commit=View+differences)
